GE21-ME21 segment reconstruction

slowmoyang opened the pull request:
#### PR description:
- This PR implements the GE21-ME21 segment reconstruction in the same way as the GE11-ME11 segment reconstruction. 
- Add `GEMCSCSegmentProducer::fillDescription`
- Add `GEMCSCCoincidenceRateAnalyzer` for the GEM-CSC coincidence rate study
- Add `runGEMCSCCoincidenceRateAnalyzer_cfg.py` and `plotGEMCSCCoincidenceRate.py` to validate `GEMCSCSegment` reconstruction codes. The instruction is shown in the PR validation part.

#### PR validation:
#### 
Because `GEMCSCSegmentProducer` is not included in the standard reconstruction sequence, this PR cannot be validated by `runTheMatrix.py`.
 Therefore, wf 42021.0 was modified and used for the validation. [The result](https://indico.cern.ch/event/1199653/contributions/5044332/attachments/2507440/4308901/GEM-DPG_220913.pdf) was presented in the GEM DPG meeting.


#### via online DQM unittest
Also, because this PR touches `DQM/GEM`, this PR was valiated with the online DQM unit test. Logs and output files are [here](https://github.com/slowmoyang/GEM-Workspace/tree/main/SW-Dev/220722_GE21-ME21-Segment-Reco/test/CMSSW_12_6_0_pre2/online-dqm-unittest).
```zsh
cmsRun ${CMSSW_BASE}/src/DQM/Integration/python/clients/gem_dqm_sourceclient-live_cfg.py unitTest=True
```

#### via `runGEMCSCCoincidenceRateAnalyzer_cfg.py` and `plotGEMCSCCoincidenceRate.py`
This PR can be validated with `runGEMCSCCoincidenceRateAnalyzer_cfg.py` and `plotGEMCSCCoincidenceRate.py`. `runGEMCSCCoincidenceRateAnalyzer_cfg.py` takes RECO-tier files as input and then runs `GEMCSCSegmentProducer` and `GEMCSCCoincidenceRateAnalyzer`. `GEMCSCCoincidenceRateAnalyzer` produces a root file containing `TTree`. Then, `plotGEMCSCCoincidenceRate.py`takes the resulting root file as input and produces efficiency plots.

Results can be found [here](https://docs.google.com/presentation/d/1Yt5L1gfkJmAKLuKQ9E37ktNNccAhiQyi1VOatnzxwug/edit?usp=sharing)

```shell
$ cmsRun runGEMCSCCoincidenceRateAnalyzer_cfg.py -o ./output.root
$ python3 plotGEMCSCCoincidenceRate.py -i ./output.root
```

The help message of `runGEMCSCCoincidenceRateAnalyzer_cfg.py` is as below. The era, global tag and geometry can be configured using arguments.

```shell
$ python3 runGEMCSCCoincidenceRateAnalyzer_cfg.py -h
sys.argv=['runGEMCSCCoincidenceRateAnalyzer_cfg.py', '-h']
['runGEMCSCCoincidenceRateAnalyzer_cfg.py'] are interpreted as arguments for cmsRun.
parsing argumnets from ['-h'].
usage: runGEMCSCCoincidenceRateAnalyzer_cfg.py [-h] [-e ERA] [-t GLOBAL_TAG] [-g GEOMETRY] [-m MAX_EVENTS] [-i INPUT_FILES [INPUT_FILES ...]] [-o OUTPUT_FILE]

optional arguments:
  -h, --help            show this help message and exit
  -e ERA, --era ERA     era (default: Phase2C17I13M9)
  -t GLOBAL_TAG, --global-tag GLOBAL_TAG
                        global tag (default: auto:phase2_realistic_T21)
  -g GEOMETRY, --geometry GEOMETRY
                        geometry (default: GeometryExtended2026D88Reco)
  -m MAX_EVENTS, --max-events MAX_EVENTS
                        max events (default: -1)
  -i INPUT_FILES [INPUT_FILES ...], --input-files INPUT_FILES [INPUT_FILES ...]
                        input files (default: ['file:/eos/cms/store/relval/CMSSW_12_6_0_pre2/RelValSingleMuPt1000/GEN-SIM-
                        RECO/125X_mcRun4_realistic_v2_2026D88noPU-v1/2580000/412ca5d0-4533-477d-ae92-411f5a532937.root', 'file:/eos/cms/store/relval/CMSSW_12_6_0_pre2/RelValSingleMuPt1000/GEN-SIM-
                        RECO/125X_mcRun4_realistic_v2_2026D88noPU-v1/2580000/817a15f8-1355-4606-a069-4e2d9e4f7939.root', 'file:/eos/cms/store/relval/CMSSW_12_6_0_pre2/RelValSingleMuPt1000/GEN-SIM-
                        RECO/125X_mcRun4_realistic_v2_2026D88noPU-v1/2580000/d1801dfb-b4e3-489b-b04c-5125cc35bc2e.root', 'file:/eos/cms/store/relval/CMSSW_12_6_0_pre2/RelValSingleMuPt1000/GEN-SIM-
                        RECO/125X_mcRun4_realistic_v2_2026D88noPU-v1/2580000/d21fa76d-6a29-4455-8476-cc04800a4b49.root', 'file:/eos/cms/store/relval/CMSSW_12_6_0_pre2/RelValSingleMuPt1000/GEN-SIM-
                        RECO/125X_mcRun4_realistic_v2_2026D88noPU-v1/2580000/f25378b1-c009-41b0-9a1b-1a043c4cc599.root'])
  -o OUTPUT_FILE, --output-file OUTPUT_FILE
                        output file (default: output.root)
```


#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:
This PR is not a backport but will be backported.

slowmoyang committed:
Implement the GE21-ME21 segment reconstruction

- Implement the GE21-ME21 segment reconstruction the same way as the GE11-ME11
segment reconstruction.
- Add `fillDescription` to `GEMCSCSegmentProducer`.
- Add `GEMCSCCoincidenceRateAnalyzer` for GEM-CSC coincidence rate study

slowmoyang commented:
@jshlee

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39441/32148

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39441/32148/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39441/32148/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

slowmoyang committed:
run code-checks and code-format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39441/32152

- This PR adds an extra 28KB to repository

cmsbuild commented:
A new Pull Request was created by  @slowmoyang (slowmoyang) for master.

It involves the following packages:

- DQM/GEM (**dqm**)
- DQM/Integration (**dqm**)
- RecoLocalMuon/GEMCSCSegment (**upgrade, reconstruction**)


@micsucmed, @emanueleusai, @ahmad3213, @cmsbuild, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @syuvivida, @pmandrik, @mandrenguyen, @rvenditti can you please review it and eventually sign? Thanks.
@batinkov, @bellan, @battibass, @watson-ij, @missirol, @jhgoh, @trtomei, @threus, @beaucero, @francescobrivio this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mandrenguyen commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-774acf/27649/summary.html
**COMMIT**: 4b893a262a2d5918c6dbc1834bfc2bc20d6bc8cb
**CMSSW**: CMSSW_12_6_X_2022-09-18-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39441/27649/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 1 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3621956
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3621929
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-18-2300+774acf/52865/triggerResults): no differences found

mandrenguyen commented:
Hi @slowmoyang  can you provide a recipe to test this PR?
I'm not even finding a relval wf with the number 42021.0 in 12_6_0_pre2.
Why do we need to have GEMCSCCoincidenceRateAnalyzer in cmssw instead of some private repo? 
Is there a reason to put it in test instead of plugins?

slowmoyang commented:
Hi @mandrenguyen,

I tested this PR using cfg files [here](https://github.com/slowmoyang/GEM-Workspace/tree/main/SW-Dev/220722_GE21-ME21-Segment-Reco/test/CMSSW_12_6_0_pre2/TenMu_phase2_realistic_T21) The step1-3 cfg files are the output of wf 42021.0 with a bit modification (removing redundant steps like `PAT` and extending output contents). Then, `step4_GEMCSCSegmentReco.py` calls `GEMRecHitProducer` with `ge21Off` of `true` and `GEMCSCSegmentProducer`. Finally, `step5_Analysis.py` calls `GEMCSCCoincidenceRateAnalyzer`, which produces `TTree` using `TFileService`.
```zsh
cmsrel CMSSW_12_6_0_pre2
cd CMSSW_12_6_0_pre2/src
cmsenv
git-cms-merge-topic slowmoyang:ge21-me21-seg-reco__from-CMSSW_12_6_0_pre2
curl -OJ https://raw.githubusercontent.com/slowmoyang/GEM-Workspace/main/SW-Dev/220722_GE21-ME21-Segment-Reco/test/CMSSW_12_6_0_pre2/TenMu_phase2_realistic_T21/step1_GEN_SIM.py
curl -OJ https://raw.githubusercontent.com/slowmoyang/GEM-Workspace/main/SW-Dev/220722_GE21-ME21-Segment-Reco/test/CMSSW_12_6_0_pre2/TenMu_phase2_realistic_T21/step2_DIGI_L1TrackTrigger_L1_DIGI2RAW_HLT.py
curl -OJ https://raw.githubusercontent.com/slowmoyang/GEM-Workspace/main/SW-Dev/220722_GE21-ME21-Segment-Reco/test/CMSSW_12_6_0_pre2/TenMu_phase2_realistic_T21/step3_RAW2DIGI_RECO.py
curl -OJ https://raw.githubusercontent.com/slowmoyang/GEM-Workspace/main/SW-Dev/220722_GE21-ME21-Segment-Reco/test/CMSSW_12_6_0_pre2/TenMu_phase2_realistic_T21/step4_GEMCSCSegmentRECO.py
curl -OJ https://raw.githubusercontent.com/slowmoyang/GEM-Workspace/main/SW-Dev/220722_GE21-ME21-Segment-Reco/test/CMSSW_12_6_0_pre2/TenMu_phase2_realistic_T21/step5_Analysis.py
cmsRun step1_GEN_SIM.py
cmsRun step2_DIGI_L1TrackTrigger_L1_DIGI2RAW_HLT.py
cmsRun step3_RAW2DIGI_RECO.py
cmsRun step4_GEMCSCSegmentRECO.py
cmsRun step5_Analysis.py
```

First of all, I'm okay to drop `RecoLocalMuon/GEMCSCSegment/test/GEMCSCCoincidenceAnalyzer.cc` because it is not used in the unit test. Anyways, I added `GEMCSCCoincidenceRateAnalyezr` to replace `RecoLocalMuon/GEMCSCSegment/test/TestGEMCSCSegmentAnalyzer.cc`. `TestGEMCSCSegmentAnalyzer` seems to be used for `GEMCSCSegmentProducer` validation and produces more than 200 histograms only for GE11-ME11 segments. It is straightforward to modify `TestGEMCSCSegmentAnalyzer` for GE21-ME21 segments. But I think it is more useful to produce `TTree` instead of histograms for validating `GEMCSCSegmentProducer`, although the current `GEMCSCCoincidenceAnalyzer` does not cover all output of `TestGEMCSCSegmentAnalyzer`.

slowmoyang commented:
Also, wf42021.0 exists in the upgrade set.
```console
$ runTheMatrix.py --what upgrade --showMatrix | grep '42021\.0'
42021.0 2026D94PU+TenMuE_0_200_pythia8_GenSimHLBeamSpot+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU
```

mandrenguyen commented:
I'm still trying to understand why this isn't tested in a relval workflow.  Is there a plan to introduce `GEMCSCSegmentProducer` into the standard reco sequence?

srimanob commented:
Hi @slowmoyang 
I see that your step 4 includes the extension of FEVTDEBUG also, is this expected in general (i.e. when run this together with RECO)?

slowmoyang commented:
@mandrenguyen As far as I know, `GEMCSCSegment` was introduced for improving muon collection and is still being studied. So, `GEMCSCSegmentProducer` was not intentionally included in the standard reco sequence because it is an experimental feature. In the view of the GEM DQM side, `GEMCSCSegment` is used for the online DQM, which is independent of standard sequences.

(I may be wrong in this, so @jshlee can correct me if I am.)

slowmoyang commented:
Hi @srimanob

Because `GEMCSCSegmentProducer` labelled as `gemcscSegmets` is not included in any event content, so the output event content is extended for the next step. In the standard sequence, there is no module requiring `GEMCSCSegment` and so we don't need the extension.

emanueleusai commented:
+1

slowmoyang commented:
@mandrenguyen, are you suggesting introducing a dedicated wf through this PR? If so, can you provide a recipe?

slowmoyang committed:
add a modifier to test GEMCSCSegmentProducer

add a modifier to test GEMCSCSegmentProducer in relval workflows, as
suggested in https://github.com/cms-sw/cmssw/pull/39441#issuecomment-1256109692

slowmoyang committed:
replace Run3_GEM with gemcscSegmentsTesting

Turn on GE21-ME21 segment reconstruction only in relval workflows.

slowmoyang commented:
@mandrenguyen I added a modifier named as `gemcscSegmentsTesting` to test `GEMCSCSegmentProducer`. Also, in `gemcscsSegments_cff.py`, I replaced `Run3_GEM` modifier , which turns on GE21-ME21 segment recon, with `gemcscSegmentsTesting`.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39441/32234

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #39441 was updated. @perrotta, @rappoccio, @pmandrik, @mandrenguyen, @emanueleusai, @ahmad3213, @cmsbuild, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @syuvivida, @rvenditti, @micsucmed, @fabiocos, @davidlange6 can you please check and sign again.

mandrenguyen commented:
> @mandrenguyen I added a modifier named as `gemcscSegmentsTesting` to test `GEMCSCSegmentProducer`. Also, in `gemcscsSegments_cff.py`, I replaced `Run3_GEM` modifier , which turns on GE21-ME21 segment recon, with `gemcscSegmentsTesting`.

I'm not sure I understand.  Was run3_GEM already giving you the behavior you want?  In that case, there's no need to introduce a process modifier.  One simply needs a relval workflow to test this era.  
Do I understand correctly that this PR is only intended for run 3 reconstruction, and will not touch Phase 2?  It would be helpful to explain that in the PR description.

slowmoyang commented:
@mandrenguyen I removed the Run3 modifier, which turns on GE21-ME21 segment reconstruction. This is because I realised that people studying the impact of GE11-ME11 segments on muons don't want GE21-M21, though the GE11-M11 study is offline. Then, I added the modifier to enable testing of GE21-ME21 segments in relval workflows. Sorry for confusing you.

emanueleusai commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-774acf/27780/summary.html
**COMMIT**: c8830e2f983d723842e0a3f57d68285204c62029
**CMSSW**: CMSSW_12_6_X_2022-09-26-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39441/27780/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3624368
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3624344
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-26-1100+774acf/52972/triggerResults): no differences found

emanueleusai commented:
+1

- Comment: this PR includes changes to a DQM online client, a backport to 12_4 is required in order to deploy the changes to DQM machines at P5

mandrenguyen commented:
+reconstruction
As discussed at today's reco meeting, this PR will not be backported to 12_4_X, even though it's currently being used for the online DQM.
Although this PR is not currently testable in relVal, the GEM folks will consider incorporating it in an era/procMod for subsequent developments. 
@slowmoyang Please also make a 12_5_X backport.

emanueleusai commented:
@mandrenguyen thanks for the update, we still need a PR to 12_4 to be able to include this in DQM online, even if it will not be merged and made part of an official 12_4 release. We can put the 12_4 backport on hold, similarly to what is being done with https://github.com/cms-sw/cmssw/pull/39489

srimanob commented:
@cmsbuild please test

Just re-trigger the test after almost a week.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-774acf/27911/summary.html
**COMMIT**: c8830e2f983d723842e0a3f57d68285204c62029
**CMSSW**: CMSSW_12_6_X_2022-10-02-0000/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39441/27911/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-774acf/41834.0_TTbar_14TeV+2026D94+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3432650
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3432622
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 49 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-10-02-0000+774acf/53097/triggerResults): no differences found

srimanob commented:
@cmsbuild please test

srimanob commented:
+Upgrade

srimanob commented:
@cmsbuild please test

To be ready for merging.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-774acf/28496/summary.html
**COMMIT**: c8830e2f983d723842e0a3f57d68285204c62029
**CMSSW**: CMSSW_12_6_X_2022-10-25-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39441/28496/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - Reco comparison had  6 failed jobs
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3378384
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3378359
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 201 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-10-25-1100+774acf/53647/triggerResults): no differences found

perrotta commented on RecoLocalMuon/GEMCSCSegment/test/GEMCSCCoincidenceRateAnalyzer.cc:
```diff
@@ -0,0 +1,356 @@
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/ESHandle.h"
```
This does not seem to be needed

perrotta commented on RecoLocalMuon/GEMCSCSegment/plugins/GEMCSCSegmentBuilder.cc:
```diff
@@ -214,13 +214,13 @@ void GEMCSCSegmentBuilder::build(const GEMRecHitCollection* recHits,
-                                               << CSCId.rawId() << " = " << CSCId << std::endl;
+    else {
+      edm::LogVerbatim("GEMCSCSegmentBuilder")
+          << "[GEMCSCSegmentBuilder :: build] Found a Segment in " << CSCId.rawId() << " = " << CSCId << std::endl;
```
```suggestion
          << "[GEMCSCSegmentBuilder :: build] Found a Segment in " << CSCId.rawId() << " = " << CSCId;
```

perrotta commented on RecoLocalMuon/GEMCSCSegment/plugins/GEMCSCSegmentProducer.h:
```diff
@@ -14,6 +14,8 @@
 #include "FWCore/Framework/interface/Event.h"
 #include "FWCore/Framework/interface/ESHandle.h"
```
Also this one is not needed (just in case)

perrotta commented on RecoLocalMuon/GEMCSCSegment/test/GEMCSCCoincidenceRateAnalyzer.cc:
```diff
@@ -0,0 +1,356 @@
+  tree_->Branch("gem_cls", &b_gem_cls_);
+}
+
+GEMCSCCoincidenceRateAnalyzer::~GEMCSCCoincidenceRateAnalyzer() {}
+
```
```suggestion
```

perrotta commented on RecoLocalMuon/GEMCSCSegment/test/GEMCSCCoincidenceRateAnalyzer.cc:
```diff
@@ -0,0 +1,356 @@
+class GEMCSCCoincidenceRateAnalyzer : public edm::one::EDAnalyzer<edm::one::SharedResources> {
+public:
+  explicit GEMCSCCoincidenceRateAnalyzer(const edm::ParameterSet&);
+  ~GEMCSCCoincidenceRateAnalyzer() override;
```
```suggestion
  ~GEMCSCCoincidenceRateAnalyzer() override = default;
```

perrotta commented on RecoLocalMuon/GEMCSCSegment/plugins/GEMCSCSegmentBuilder.cc:
```diff
@@ -120,14 +121,13 @@ void GEMCSCSegmentBuilder::build(const GEMRecHitCollection* recHits,
       edm::LogVerbatim("GEMCSCSegmentBuilder")
-          << "[GEMCSCSegmentBuilder :: build] Found ME1/1 Segment in " << CSCId.rawId() << " = " << CSCId << std::endl;
+          << "[GEMCSCSegmentBuilder :: build] Found " << (CSCId.isME11() ? "ME1/1" : "ME2/1") << " Segment in "
+          << CSCId.rawId() << " = " << CSCId << std::endl;
```
```suggestion
          << CSCId.rawId() << " = " << CSCId;
```

perrotta commented:
@slowmoyang (and @emanueleusai @mandrenguyen) maybe I run the whole thread a bot too fast, but just to understand:
- if this new feature is not testable with standard workflows (also data workflows) why was it ported to online DQM?
- could you please confirm that the tests run fine even after last changes (modifier, etc) and provide a recipe to replicate them?
- could you please edit the PR description and update it with all latest developments?

slowmoyang committed:
accept perrotta's suggestions

Co-authored-by: Andrea Perrotta <perrotta@cern.ch>

The following include directives are kept in
`RecoLocalMuon/GEMCSCSegment/test/GEMCSCCoincidenceRateAnalyzer.cc`:
`#include "FWCore/Framework/interface/Frameworkfwd.h`
`#include "FWCore/Framework/interface/one/EDAnalyzer.h`

slowmoyang commented:
@perrotta Thanks for your review. I will make updates and reply to your questions within a few days.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39441/32822

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #39441 was updated. @perrotta, @rappoccio, @pmandrik, @mandrenguyen, @emanueleusai, @ahmad3213, @cmsbuild, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @syuvivida, @rvenditti, @micsucmed, @fabiocos, @davidlange6 can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39441/32967

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #39441 was updated. @perrotta, @rappoccio, @pmandrik, @mandrenguyen, @emanueleusai, @ahmad3213, @cmsbuild, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @syuvivida, @rvenditti, @micsucmed, @fabiocos, @davidlange6 can you please check and sign again.

slowmoyang commented:
I just accepted suggestions but got compilation. I will quickly fix it.

slowmoyang committed:
add files to validate GEMCSCSegmentProducer

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39441/32992

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/Integration/python/clients/gem_dqm_sourceclient-live_cfg.py  modified in PR(s): #40034

cmsbuild commented:
Pull request #39441 was updated. @perrotta, @rappoccio, @pmandrik, @mandrenguyen, @emanueleusai, @ahmad3213, @cmsbuild, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @syuvivida, @rvenditti, @micsucmed, @fabiocos, @davidlange6 can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39441/32993

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQM/Integration/python/clients/gem_dqm_sourceclient-live_cfg.py  modified in PR(s): #40034

slowmoyang commented:
@perrotta

> if this new feature is not testable with standard workflows (also data workflows) why was it ported to online DQM?

GEMCSCSegment is being used in [the online DQM for monitoring GEM efficiency](https://github.com/cms-sw/cmssw/blob/master/DQM/Integration/python/clients/gem_dqm_sourceclient-live_cfg.py#L65), and this PR implements GE21-ME21 segment reconstruction for upcoming GE21 monitoring.


> could you please confirm that the tests run fine even after last changes (modifier, etc) and provide a recipe to replicate them?
Because I removed files used in the original test, I tested current version of codes with RelVal data. Then, results looks consistent.

I added files for validation to this PR. The instruction can be found in the PR description. 
Since files used in the original test have been removed, I tested with the current version of codes using relval data. The new validation results look consistent with the original one.

> could you please edit the PR description and update it with all latest developments?

I updated it.

cmsbuild commented:
Pull request #39441 was updated. @perrotta, @rappoccio, @pmandrik, @mandrenguyen, @emanueleusai, @ahmad3213, @cmsbuild, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @syuvivida, @rvenditti, @micsucmed, @fabiocos, @davidlange6 can you please check and sign again.

emanueleusai commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-774acf/28979/summary.html
**COMMIT**: 447d0398758c97cb9c6c5e209b8f9284a2df7dbd
**CMSSW**: CMSSW_12_6_X_2022-11-11-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39441/28979/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416993
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416971
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-11-2300+774acf/54059/triggerResults): no differences found

emanueleusai commented:
+1

- resign

srimanob commented:
+Upgrade
re-sign

perrotta commented:
+1
- PR tests are a bit old, but this PR shouldn't touch anything in the normal workflows

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
